### PR TITLE
Plugin save and load.

### PR DESCRIPTION
**Pull request by anusha-ragunathan**
Implement plugin save and load to handle the offline distribution
usecase. Save a tar of the plugin in the following layout:
--- sha256ofconfig.json
--- sha256ofrootfs.tar.gz
--- manifest.json containing metadata of the form:
{
	"Name"  : pluginName
	"Config": sha256config.json
	"Layers": [
		"sha256rootfs.tar.gz"
	]
}
Load loads such a tar stream into the plugin inventory.

Signed-off-by: Anusha Ragunathan <anusha.ragunathan@docker.com>

### CODE REVIEW

File: plugin/manager.go [hunk 1]
```diff
@@ -161,6 +161,47 @@ func (pm *Manager) StateChanged(id string, e libcontainerd.StateInfo) error {
...
+			if p.PluginObj.Config.PropagatedMount != "" {
+				propRoot := filepath.Join(filepath.Dir(p.Rootfs), "propagated-mount")
+
+				// check if we need to migrate an older propagated mount from before
+				// these mounts were stored outside the plugin rootfs
+				if _, err := os.Stat(propRoot); os.IsNotExist(err) {
+					if _, err := os.Stat(p.PropagatedMount); err == nil {
+						// make sure nothing is mounted here
+						// don't care about errors
+						mount.Unmount(p.PropagatedMount)
```
Comment (cpuguy83) [inline plugin/manager.go]: Recursive unmount?

File: cli/command/plugin/load.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		Args:  cli.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runLoad(dockerCli, &opts)
+		},
+	}
+
+	flags := cmd.Flags()
+
+	flags.StringVarP(&opts.input, "input", "i", "", "Read from tar archive file, instead of STDIN")
+	// XXX: quiet for plugins seems unnecessary. Nothing verbose gets printed out anyways.
```
Comment (cpuguy83) [inline cli/command/plugin/load.go]: Maybe we should remove for now?

File: plugin/backend_linux.go [hunk 3]
```diff
@@ -797,6 +813,262 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+	}
+
+	p.PluginObj.PluginReference = manifest.Name
+
+	pm.config.LogPluginEvent(p.PluginObj.ID, manifest.Name, "load")
+
+	outStream.Write([]byte(fmt.Sprintf("Loaded plugin ID: %s\n", p.PluginObj.ID)))
+
+	// call mountPropUpdate to finalize mount magic for volume plugins.
+	mountPropUpdate(p)
```
Comment (cpuguy83) [inline plugin/backend_linux.go]: Wouldn't this only need to be called when the plugin is enabled?

**Review by cpuguy83 — DISMISSED** [review]
Design LGTM

A couple of code comments.

File: plugin/backend_linux.go [hunk 4]
```diff
@@ -797,6 +813,262 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+
+	// XXX: Instead of using the blobStore interface to create, write and commit
+	// to blobstore, why not simply copy the files and rename them?
+
+	// write rootfs to blobStore
+	rootfsBlob, err := pm.blobStore.New()
+	if err != nil {
+		return err
+	}
+	rootfsRdr := io.TeeReader(compressedTarRdr, rootfsBlob)
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: No need for `TeeReader`. Just `io.Copy`

File: plugin/backend_linux.go [hunk 5]
```diff
@@ -797,6 +813,262 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+	if err != nil {
+		return err
+	}
+	// create a tmpdir to extract the decompressedTarRdr
+	tmpRootFS, err := ioutil.TempDir(tmpDir, ".rootfs")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpRootFS)
+	if err = chrootarchive.Untar(decompressedTarRdr, tmpRootFS, nil); err != nil {
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: `ApplyLayer` needs to be used here. Could probably also combine into utility function to share this code with install.

File: plugin/manager_linux.go [hunk 6]
```diff
@@ -238,56 +238,58 @@ func (pm *Manager) upgradePlugin(p *v2.Plugin, configDigest digest.Digest, blobs
...
-func (pm *Manager) createPlugin(name string, configDigest digest.Digest, blobsums []digest.Digest, rootFSDir string, privileges *types.PluginPrivileges) (p *v2.Plugin, err error) {
+func (pm *Manager) createPlugin(name string, configDigest digest.Digest, blobsums []digest.Digest, rootFSDir string, privileges *types.PluginPrivileges, config *types.PluginConfig) (p *v2.Plugin, err error) {
 	if err := pm.config.Store.validateName(name); err != nil { // todo: this check is wrong. remove store
 		return nil, err
 	}
 
-	config, err := pm.setupNewPlugin(configDigest, blobsums, privileges)
-	if err != nil {
-		return nil, err
+	if config == nil {
```
Comment (tonistiigi) [inline plugin/manager_linux.go]: Don't quite understand why this change was needed. Config is still in the blobstore when loaded, right?

File: plugin/backend_linux.go [hunk 7]
```diff
@@ -797,6 +813,262 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+//		"Name"  : pluginName
+//		"Config": sha256config.json
+//		"Layers": [
+//			"sha256rootfs.tar.gz"
+//		]
+//	}
+/* XXX: Eventually add a OCI image spec complaint layout
+ * https://github.com/opencontainers/image-spec/blob/master/image-layout.md
+ */
+func (pm *Manager) SavePlugin(plugin string, outStream *ioutils.WriteFlusher) error {
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: why is this `WriteFlusher`?

**Comment by anusha-ragunathan** [comment]
Got bit by 34337db95d8eb85875509e341f452d8b555452b8.

**Review by tonistiigi — COMMENTED** [review]
I'm bit hesitant to introduce a format that is very similar to OCI format. They look very similar so we could avoid a need to write yet another spec. This isn't something that can be easily changed later. Is there some issues that make switching to that format harder.

I think the model where you specify a name on `plugin load` may work better for plugins as names are descriptions of plugin instances. Especially if grant settings need to specified by user as well. One problem with that may be that oci format defines multiple manifests, but I guess we could limit plugin loading to single manifest.

**Comment by anusha-ragunathan** [comment]
https://github.com/docker/cli/pull/29 for the cli/client changes. Will remove those parts from this PR and add vendor soon.

**Comment by anusha-ragunathan** [comment]
@cpuguy83 will be carrying the PR going forward. Thank you!

**Comment by mlaventure** [comment]
ping @cpuguy83 :angel:

**Comment by cpuguy83** [comment]
I'm on it, but had to spend some time going over the OCI spec first.

File: plugin/backend_linux.go [hunk 8]
```diff
@@ -797,6 +810,318 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+			MediaType: mediaType,
+		})
+
+		if mediaType == schema2.MediaTypePluginConfig {
+			manifest.Config.Size = size
+		}
+	}
+
+	manifest.Config.MediaType = mediaTypePluginConfig
+	manifest.Config.Annotations = map[string]string{"com.docker.plugin.ref.name": srcP.PluginObj.Name}
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: These should be part of index. https://github.com/opencontainers/image-spec/blob/master/annotations.md#pre-defined-annotation-keys

File: plugin/backend_linux.go [hunk 9]
```diff
@@ -797,6 +810,318 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+	tempDir, err := ioutil.TempDir("", "docker-plugin-export-")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+
+	// 2. save config and rootfs into tmpdir
+	var manifest ocispecv1.Manifest
+
+	for i, dgst := range []digest.Digest{srcP.Config, srcP.Blobsums[0]} {
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: Handle all the blobs, not only first

File: plugin/backend_linux.go [hunk 10]
```diff
@@ -797,6 +810,318 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+			return errors.Wrap(err, "error writing blob")
+		}
+
+		// set access and modified times of the manifest to be unix epoch, so that
+		// tar checksum is consistent.
+		if err := system.Chtimes(dest, time.Unix(0, 0), time.Unix(0, 0)); err != nil {
+			return errors.Wrap(err, "error resetting mod times for blob")
+		}
+
+		manifest.Layers = append(manifest.Layers, ocispecv1.Descriptor{
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: Config should not go to layers.

File: plugin/backend_linux.go [hunk 11]
```diff
@@ -797,6 +810,318 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+		rdr, err := pm.blobStore.Get(dgst)
+		if err != nil {
+			return err
+		}
+		defer rdr.Close()
+
+		var mediaType string
+		if i == 0 {
+			mediaType = schema2.MediaTypePluginConfig
+			manifest.Config.Digest = dgst
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: Change config mediatype

File: plugin/backend_linux.go (hunk 8 above)
Comment (stevvooe) [inline plugin/backend_linux.go]: This is for tags and isn't really needed here.

Comment (stevvooe) [inline plugin/backend_linux.go]: Wait, I misunderstood this. Do what tonis said, put it on the index.

File: plugin/backend_linux.go [hunk 12]
```diff
@@ -797,6 +810,318 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
 	return nil
 }
 
+// SavePlugin saves the plugin in outStream
+// The tar layout is as follows:
+// --- sha256ofconfig.json
```
Comment (stevvooe) [inline plugin/backend_linux.go]: Are we still doing this format? It is a little odd...

Comment (tonistiigi) [inline plugin/backend_linux.go]: No, this comment is from pre-oci commit.

File: plugin/backend_linux.go (hunk 8 above)
Comment (tonistiigi) [inline plugin/backend_linux.go]: Actually, it's a bit weird as you do want to annotate a specific manifest, not all of them. Currently, we don't have the issue as you can't have multiple manifests but not sure how it is supposed to work with images. You can't set it on manifest though as manifests need to be immutable.

Comment (cpuguy83) [inline plugin/backend_linux.go]: Looking at https://github.com/moby/moby/pull/26369/files#diff-23f8e73badb7caf4300fe141650c2880R92
It looks like he was putting this annotation on the manifest descriptor in the index.json, which seems reasonable.

Comment (tonistiigi) [inline plugin/backend_linux.go]: Ah, yes. Of course. I was confused by the example in https://github.com/opencontainers/image-spec/blob/master/image-index.md#example-image-index

Comment (stevvooe) [inline plugin/backend_linux.go]: @tonistiigi The index is supposed to allow you to tag different manifests as tags with ref.name. For example, you could have `org.opencontianers.image.ref.name=latest` in index.json would be the latest tag.

I can work up a better example, if that helps. We may want to get something more illustrative in the spec, but https://github.com/opencontainers/image-spec/blob/master/image-layout.md#index-example shows somewhat how this should work.

Comment (cpuguy83) [inline plugin/backend_linux.go]: Ok, I've at least moved this over to the manifest descriptor.

Made other fixes and some cleanup as well.

**Comment by cpuguy83** [comment]
I've updated this to use the OCI image spec for save/load.
PTAL, especially that the implementation of the spec is correct.

It seems like there's still some question around how we should handle naming within the confines of the OCI spec. Currently using annotations to specify the plugin name at least for now (@stevvooe).

I've intentionally not rebased while it's in design review.

File: plugin/backend_linux.go [hunk 13]
```diff
@@ -797,6 +808,251 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+
+			size, err := io.Copy(rootfsBlob, compressedTar)
+			if err != nil {
+				return errors.Wrap(err, "error reading layer")
+			}
+			if size != l.Size {
+				return errors.New("layer size mismatch")
+			}
+
+			if _, err := rootfsBlob.Commit(); err != nil {
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: This returns digest that can be verified

File: plugin/backend_linux.go [hunk 14]
```diff
@@ -797,6 +808,251 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+
+			// decompress for plugin rootfs
+			if _, err := compressedTar.Seek(0, 0); err != nil {
+				return errors.Wrap(err, "error seeking to beginning of layer stream")
+			}
+			decompressedTar, err := archive.DecompressStream(compressedTar)
+			if err != nil {
+				return err
+			}
+			if err = chrootarchive.Untar(decompressedTar, tmpRootFS, nil); err != nil {
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: `chrootarchive.ApplyLayer`

File: plugin/backend_linux.go [hunk 15]
```diff
@@ -797,6 +808,251 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
...
+			if size != l.Size {
+				return errors.New("layer size mismatch")
+			}
+
+			if _, err := rootfsBlob.Commit(); err != nil {
+				return err
+			}
+
+			// decompress for plugin rootfs
+			if _, err := compressedTar.Seek(0, 0); err != nil {
```
Comment (tonistiigi) [inline plugin/backend_linux.go]: This can be done with `TeeReader` instead of reading twice. Look at the pull code, or even better to extract this into same function that both pull and load use https://github.com/moby/moby/blob/master/plugin/blobstore.go#L141-L153

Should also verify the blob against the config. https://github.com/moby/moby/blob/master/plugin/blobstore.go#L149

**Comment by cpuguy83** [comment]
Code is still far from where I want it to be, but this is doing full verification of the loaded plugin image.

**Review by tonistiigi — COMMENTED** [review]
OCI parts seem ok to me. Added couple of comments and some previous ones still seem valid.

Not sure what's the best approach for the privileges check. The comment says that it should be exposed by the user on load and I agree with that. Again, this doesn't seem to be something that can be easily changed later. Maybe just force users to send `--grant-all-privileges` for now?

**Comment by cpuguy83** [comment]
@tonistiigi One thing we could do is have the client send the list of privileges to grant in order that they are in the manifest.
This will require the client to inspect the bundle, but I think that's ok.

**Comment by tonistiigi** [comment]
@cpuguy83 You mean doing it when the user specifies `--grant-all` or would that just be sent to daemon as boolean? Could probably also just fail on loading and make the user retry.

**Comment by cpuguy83** [comment]
@tonistiigi 

```
$ docker plugin load -i foo.tar
<client reads tar>
plugin foo1 requests these perms: y/N?
<perms>
plugin foo2 requests these perms: y/N?
<perms>

POST /plugins/load
{
    [         
        { "privileges": <computed foo1 privs> },
        { "privileges": <computed foo2 privs> }
    ]
}
```

Or something like that...

**Comment by cpuguy83** [comment]
I think I've addressed all comments except how to deal with privileges.

**Comment by cpuguy83** [comment]
Maybe we can just stream two objects, first the privileges and then the tar?

**Comment by thaJeztah** [comment]
ping @tonistiigi @stevvooe PTAL

**Comment by stevvooe** [comment]
@thaJeztah Is there a description of the current plugin format that this PR implements?

**Comment by thaJeztah** [comment]
@stevvooe if there is, it's probably in this directory; https://github.com/docker/cli/tree/master/docs/extend

**Comment by stevvooe** [comment]
@thaJeztah This is an engineering problem. Where is the format specification? This needs to go through review and be carefully considered, as it cannot be changed once we move forward.

**Comment by AkihiroSuda** [comment]
ping @anusha-ragunathan

**Comment by thaJeztah** [comment]
We'll discuss this again when @anusha-ragunathan is back from maternity leave next month

**Comment by anusha-ragunathan** [comment]
@cpuguy83 I think we should reopen this one and complete the PR. Thoughts?

**Comment by cpuguy83** [comment]
Yeah, we need it but there's still some issues to resolve.

**Comment by thaJeztah** [comment]
hope you kept the branch somewhere 😅, otherwise here's the patch from GitHub cache:
[33032.patch.txt](https://github.com/moby/moby/files/1732864/33032.patch.txt)
